src/python/gudhi/datasets/remote.py

```diff
@@ -0,0 +1,103 @@
+#
+# Copyright (C) 2021 Inria
+#
+# Modification(s):
+#   - YYYY/MM Author: Description of the modification
+
+import hashlib
+
+from os.path import join, exists
+from os import makedirs
+
+from urllib.request import urlretrieve
+
+
+def _checksum_sha256(file_path):
+    """
+    Compute the file checksum using sha256
+
+    Parameters
+    ----------
+    file_path: string
+        Full path of the created file.
+
+    Returns
+    -------
+        The hex digest of file_path
+    """
+    sha256_hash = hashlib.sha256()
+    chunk_size = 4096
+    with open(file_path,"rb") as f:
+        # Read and update hash string value in blocks of 4K
+        while True:
+            buffer = f.read(chunk_size)
+            if not buffer:
+                break
+            sha256_hash.update(buffer)
+    return sha256_hash.hexdigest()
+
+def fetch(url, filename, dirname = "remote_datasets", checksum_flag = False, file_checksum = None):
+    """
+    Fetch the wanted dataset from the given url and save it in file_path
+
+    Parameters
+    ----------
+    url : string
+        The url to fetch the dataset from.
+    filename : string
+        The name to give to downloaded file.
+    dirname : string
+        The directory to save the file to. Default is "remote_datasets".
+    checksum_flag : boolean
+        To set if the user wants the file checksum. Default is 'False'.
+        Note that if checksum_flag is set to 'True', the file_checksum must be provided.
+    file_checksum : string
+        The file checksum using sha256 to check against the one computed on the downloaded file.
+        To be considered, checksum_flag must be set to 'True'.
+        Default is 'None'.
+
+    Returns
+    -------
+    file_path: string
+        Full path of the created file.
+    """
+    if not exists(dirname):
+        makedirs(dirname)
+
+    file_path = join(dirname, filename)
+
+    urlretrieve(url, file_path)
+
+    if (checksum_flag):
+        if file_checksum is None:
+            raise ValueError("The file checksum must be provided - different from None - for the check to be performed.")
+
+        checksum = _checksum_sha256(file_path)
+        if file_checksum != checksum:
+            raise IOError("{} has a SHA256 checksum : {}, "
+                          "different from expected : {}."
+                          "The file may be corrupted or the given url may be wrong !".format(file_path, checksum,
+                                                      file_checksum))
+
+    return file_path
+
+def fetch_spiral_2d(filename = "spiral_2d.csv", dirname = "remote_datasets"):
+    """
+    Fetch spiral_2d.csv remotely
+
+    Parameters
+    ----------
+    filename : string
+        The name to give to downloaded file. Default is "spiral_2d.csv"
+    dirname : string
+        The directory to save the file to. Default is "remote_datasets".
+
+    Returns
+    -------
+    file_path: string
+        Full path of the created file.
```

Comment:
I wonder about that. We can call fetch_spiral_2d then some readcsv function on that file. But for a user, it may be more convenient if fetch_spiral_2d already returns a numpy array of shape (n,2) (or maybe a dictionary containing this array and a few other things). Then the fact that there is some CSV file becomes an implementation detail, and we can switch to some binary format later to save space and bandwidth.

Reply:
If we do that, the user wouldn't be able to use the raw fetched files as inputs to other functions/utilities in Gudhi. Unless we provide two functions : 

- Read data from the file directly without downloading it, and return an array (or a dictionary ...).
- Download the file and return its corresponding path.

Or we keep the `fetch_spiral_2d` function for users who want an array or equivalent, and the `fetch` for those who want the file itself and make them provide the url (and checksum if wanted) ?

There is also different types of files (csv, off, pers, txt, fsc) which would make the data processing different in the first case.

Reply:
> the user wouldn't be able to use the raw fetched files as inputs to other functions/utilities in Gudhi

That's true for utilities. Although if you are using utilities, starting python to fetch the file is probably not much easier than downloading it directly (or doing a checkout of gudhi-data).
For python, that actually looks like a good thing. You shouldn't pass files directly to anything except the functions read*, the possibility to pass files to AlphaComplex for instance is something that we are deprecating.

> There is also different types of files (csv, off, pers, txt, fsc) which would make the data processing different in the first case.

In most cases a user doesn't care how we encoded our file, all they want are the coordinates (the array), and it is actually easier for them if we abstract it away from them.

(I know I didn't answer the questions :wink: )

Reply:
I think what I was describing roughly matches what sklearn does, for instance in https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_openml.html . The fetch_thing function returns (a dictionary containing) an array, which is what the user wants for immediate use. If they want to, they can use numpy.savetxt to save it as a CSV file. We could easily provide a `write_points_to_off_file` to go with the read variant so users can still get an OFF file if they want. Sklearn then relies on joblib.Memory for caching (which I think essentially pickles the array/dict and saves it by default somewhere below `~/scikit_learn_data`). This would also make the format of the downloaded file less relevant and allow us to use some more compact compressed binary format (numpy's NPY/NPZ format?).
Indeed doing this is more work than making the function equivalent to a call to wget or curl, it means that to get a different type of file we will need slightly different code to parse it. I don't think we will use 10 different formats though, so it shouldn't be too much work.
I am going to ask on mattermost if others have an opinion.

Reply:
Would we want to keep the off, csv... files in the data repo for fetching or do we convert everything to binary/compressed formats npy, npz for optimized storage...?
Also as you said, sklearn uses joblib and after discussing it with @VincentRouvreau, it seems that it's preferable to stick with basic libs like numpy and not add a third party library...